### PR TITLE
Simplify the template by keeping two approaches and 3 pipelines.

### PR DESCRIPTION
**Pull request by maye-msft**
# Type of PR

- Documentation changes
- Code changes
- Test changes
- CI-CD changes

## Purpose
Simplify the template by keeping two approaches and 3 pipelines.

## Validation steps
Follow the readme to test the template with the two approach.

### CODE REVIEW

**Review by namitms — APPROVED** [review]
Looks good to me

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 1]
```diff
@@ -166,53 +273,128 @@ cd ..
...
+- **staging branch** is for intergation testing
+- **production branch** is for production deployment
+
+You can find the [document](https://docs.microsoft.com/en-us/azure/devops/repos/git/branch-policies?view=azure-devops) to set branch policy.
+
+#### 2.4.2 DevOps pipeline setup
+
+In this repo, there are several yml files which are the pipelines to support the CI/CD. you need to import the yml as build pipeline.
+
+- Import [./doveops/notebook-pipelines.yml](./devops/notebook-pipelines.yml) as build pipeline. This pipeline tests and uploads the notebooks to databricks workspace.
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:299]: nit: typo, doveops.

This is also in other places -- copy paste error? :)

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 2]
```diff
@@ -6,44 +6,49 @@
...
+When building a project in databricks, we can start from a notebook and implement the business logic in Python or SparkSQL. Before go-production, we need to create CI/CD pipelines. To reduce the effort of build CI/CD pipelines, we build this git repository as template including sample noteboks and unit tests with CI/CD pipelines in Azure DevOps yaml files .
 
 **It is the scaffolding of Azure databricks project.**
 
-To make it easy extendable, the notebooks and python code only contain super simple logic, and the unit testing is implemented by pytest and [nutter](https://github.com/microsoft/nutter)
+To make it easy extendable, the notebooks and python code only contain super simple logic, and the unit tests are implemented by pytest and [nutter](https://github.com/microsoft/nutter)
 
-This template focuses on automating provisioning, CI/CD pipeline, and various approaches of spark application implementation.
+This template focuses on solutions of CI/CD pipeline, and demostrate to support 2 approaches of spark application implementation, "notebook job" and "spark python job". A python package is also implemented and imported to notebooks and spark python job as library.
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:35]: nit: typo - demostrate

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 3]
```diff
@@ -59,105 +64,207 @@ The following technologies are used to build this template:
...
+3. [Azure account](https://portal.azure.com)
+4. [Azure Databricks Workspace](https://azure.microsoft.com/en-us/services/databricks/)
 
-   - Active subscription with the following [resource providers](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-services-resource-providers) enabled:
-     - Microsoft.Databricks
-3. [DevOps for Azure Databricks](https://marketplace.visualstudio.com/items?itemName=riserrad.azdo-databricks) It is the plugin of DevOps to support Azure Databricks deployments.
+### 2.2. Infrastructure as Code (IaC)
 
-### 2.2. Setup and deployment
+You need have 3 databricks workspace for 'develop', 'staging' and 'prodoct'. You setup Azure Databricks services by the IaC samples from [here](https://github.com/Azure-Samples/modern-data-warehouse-dataops/tree/master/single_tech_samples/databricks)
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:85]: nit: typo - "prod**o**ct" -- should be **produ**ction

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 4]
```diff
@@ -59,105 +64,207 @@ The following technologies are used to build this template:
...
-#### How to use the template
+    ```python
+    dbutils.notebook.run('../main_notebook_sql', 600)  
+    ```
 
-1. Here is a shell script to setup your project with the template
+#### 2.4.1 Repository setup
 
-```shell
+The bash script below is to ccreate a standalone git repository. You need to [create a project in Azure DevOps and create a reporsitory in the project](https://docs.microsoft.com/en-us/azure/devops/organizations/projects/create-project?view=azure-devops&tabs=preview-page). And replace the **[your repo url]** in the code below with [your reporsitory url](https://docs.microsoft.com/en-us/azure/devops/user-guide/code-with-git?toc=%2Fazure%2Fdevops%2Frepos%2Ftoc.json&bc=%2Fazure%2Fdevops%2Frepos%2Fbreadcrumb%2Ftoc.json&view=azure-devops).
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:265]: nit: typo - "ccreate" and "reporsitory"

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 5]
```diff
@@ -166,53 +273,128 @@ cd ..
...
 
-1. In your Azure DevOps project
-    - create a ci pipeline and select ci-pipeline.yml. it will run unit tests and publish the artifacts when changes committed.
-    - create a cd pipeline and select cd-pipeline.yml. it will run integration tests when ci pipeline successfully run and the go-production deployment can be manually triggered.
+After running the scripts, you can open the your repo url to check the code is pushed to the repository.
+
+There are 3 branch in the repository:
+
+- **develop branch** is the code base of development
+- **staging branch** is for intergation testing
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:290]: nit: typo - intergation

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 6]
```diff
@@ -166,53 +273,128 @@ cd ..
...
+  - Databricks-stg-environment
+  - Databricks-prod-environment
+
+  Each variable group has 3 variables:
+
+  - **databricksClusterId**: the id of Databricks cluster.
+  - **databricksDomain**: the url of Databricks workspace.
+  - **databricksToken**: the access token of Databricks.
+
+> Here are the [document](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/variable-groups?view=azure-devops&tabs=classic) of how to create variable groups, and the [docuemnt](https://docs.databricks.com/dev-tools/api/latest/authentication.html#generate-a-personal-access-token) of how to get the token.
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:320]: nit: typo - "docuemnt"

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 7]
```diff
@@ -166,53 +273,128 @@ cd ..
...
 
-1. For notebooks based spark application, you can import notebooks samples into databricks workspace from the repo of your Azure DevOps projects. Then use Databricks to edit teh notebooks. Here is an instructions how to import it.
-[https://docs.microsoft.com/en-us/azure/databricks/repos](https://docs.microsoft.com/en-us/azure/databricks/repos)
+#### 2.4.3 Import into DEV databricks workspace
 
-1. For python files based spark application, you can clone it to your local repo from the repo of your Azure DevOps projects, and then edit in your favorite IDE.
+Follow this [document](https://docs.microsoft.com/en-us/azure/databricks/repos) you can import the notebooks from the repository to databricks worksapce.
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:324]: nit: typo - "worksapce"

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 8]
```diff
@@ -166,53 +273,128 @@ cd ..
...
+- Complete the merge, it triggers the pipeline to run tests at staging databricks cluster.
 
-This folder includes a sample of notebook "main_notebook.py" which use spark sql, plus anothernotebook as its integration test.
+  ![notebook-staging](images/notebook-staging.png "notebook-staging")
 
-[Please check the detail of the sample code.](notebook-sparksql/README.md)
+#### 2.4.6 Deployment
 
-#### 2.4.3 notebook by multiple notebooks
+- Create a pull request from **staging branch** to **production branch** or directly run the pipeine on release branch.
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:346]: nit: typo - "pipeine"

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 9]
```diff
@@ -166,53 +273,128 @@ cd ..
...
+- Complete the merge, it will trigger the pipeline to run tests at staging databricks cluster.
 
-[Please check the detail of the sample code.](notebook-python-lib/README.md)
+  ![spark-python-staging](images/spark-python-staging.png "spark-python-staging")
 
-#### 2.4.5 python spark job
+#### 2.5.5 Deployment
 
-This folder includes 3 python files, the python files will be build as a spark job and library. There are 1 integration test and 2 pytest based unit tests.
+- Create a pull request from **staging branch** to **production branch** or directly run the pipeine
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:396]: nit: typo -  "pipeine"

File: single_tech_samples/databricks/sample4_ci_cd/README.md [hunk 10]
```diff
@@ -6,44 +6,49 @@
...
-  - [2.4. Sample project Structure](#24-sample-project-structure)
-    - [2.4.1 multiple notebooks](#241-multiple-notebooks)
-    - [2.4.2 notebook by pyspark API](#242-notebook-by-pyspark-API)
-    - [2.4.3 notebook by sparksql](#243-notebook-by-sparksql)
-    - [2.4.4 notebook plus python module](#244-notebook-plus-python-module)
-    - [2.4.5 python spark jo](#245-python-spark-job)
+  - [2.2. Infrastructure as Code](#22-Infrastructure-as-Code)
+  - [2.3. Project Structure](#23-Project-Structure)
+  - [2.4. The Notebook Approach](#24-The-Notebook-Approach)
+    - [2.4.1 Reposiory setup](#241-multiple-notebooks)
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/README.md:14]: nit: typo - "Reposiory"

**Review by devlace — CHANGES_REQUESTED** [review]
Some minor typos :) I'm going to do a second pass with actually testing out the sample.

File: single_tech_samples/databricks/sample4_ci_cd/README.md (hunk 1 above)
Comment (maye-msft) [reply single_tech_samples/databricks/sample4_ci_cd/README.md]: fix in [this commit](https://github.com/Azure-Samples/modern-data-warehouse-dataops/pull/359/commits/431a918f38a48c319a7a146c29bbf129fe870b46)
Thanks!

**Comment by maye-msft** [comment]
> Some minor typos :) I'm going to do a second pass with actually testing out the sample.

fix in [this commit](https://github.com/Azure-Samples/modern-data-warehouse-dataops/pull/359/commits/431a918f38a48c319a7a146c29bbf129fe870b46)
I need to run  a spell checking before next PR.
Thanks!

**Review by thurstonchen — APPROVED** [review]

**Review by devlace — APPROVED** [review]

**Pull request merged**